### PR TITLE
Make snapshot timing and trailing logs hot-reloadable in raft

### PR DESCRIPTION
**Pull request by banks**
These parameters are reasonably easy to make reloadable in a running raft instance.

Other parameters like the timeouts may be too although some require more thought about whether they are safe to change at runtime, however I'm sticking with these three right now as they are the critical parameters needed to get out of an otherwise unrecoverable situation detailed in https://github.com/hashicorp/consul/issues/9609.

There are other possible mitigations and solutions to that problem including my previous WIP PR that prevents snapshotting during replication entirely, but this has fewer uncertainties about safety and is the simplest change that gives a manual intervention option where almost none exists today.

If others are happy with this change, we can merge it and then allow Consul (and others) to have a way to re-configure snapshotting and log pruning on a leader without restarting it and loosing leadership.

## Open Questions

The API/Config change was the main gross part here. I went with this as the least-bad of a few bad options but I'm open to input if others thing there is a preferable way.

I didn't want to accept a whole `Config` even though some fields are not reloadable because then you either have to explicitly copy only some fields, or require the user ensure the rest of the config is identical to the current config (which they might not know as they can't read it) otherwise Bad Things would happen.

I also didn't want to make a separate `SetX` method for every field we make reloadable as I could see why at least a handful of others might be good to be reloadable later.

I also didn't want the internal state to continue reading from a plane `conf Config` field but just have to remember not to read reloadable fields from there but somewhere else because that seems error prone. This way forces all readers to access the most recent config and hopefully helps them consider whether the thing they are reading is reloadable or not so they can make sure the logic takes that into account (i.e. use same value in multiple operations in case it changes between them).

### CODE REVIEW

**Comment by banks** [comment]
We are still having some issues with tests in CI, but the suite passes locally for me and the new tests I added pass using `-race` too FWIW.

File: config.go
Comment (banks) [inline config.go:163]: TODO:

- [ ]  fix this comment and docs generally if we are happy with this API approach (duplicating fields)

File: api.go
Comment (banks) [inline api.go:639]:
```suggestion
// ReloadConfig updates the configuration of a running raft node. Only
// Config fields explicitly marked as reloadable will take effect.
```

Comment (banks) [inline api.go:639]:
```suggestion
// ReloadConfig updates the configuration of a running raft node. If the new config would be
// invalid it will return an error without changing the running node.
```

File: config.go
Comment (banks) [reply config.go:163]:
```suggestion
	// This can be tuned during operation using ReloadConfig.
```

**Review by briankassouf — APPROVED** [review]
One optional comment, but looks good otherwise!

File: config.go
Comment (s-christoff) [inline config.go:234]: I feel like this isn't finished here, but I may be misunderstanding it!

File: snapshot.go
Comment (s-christoff) [inline snapshot.go:223]: *Use

File: config.go
Comment (s-christoff) [reply config.go:184]: Stated below, 
" We choose to duplicate fields over embedding or accepting a Config but only using specific fields to keep the API clear."
I'm not entirely sold on the idea of embedding the ReloadableConfig into the "regular" Config because the separation of "this changes, and this doesn't" is clear and "feels safer" (which is probably not a good reason, I admit)

**Review by s-christoff — APPROVED** [review]
Some minor nits

File: config.go
Comment (banks) [reply config.go:184]: I was on the fence there - I wasn't sure if it was technically binary compatible to embed which tipped the balance although in practice most reasonable usages would probably have been unaffected. I tend to agree with Sarah though - the separation ends up being kind of OK with me in the end for keeping things clearly delineated?

File: config.go
Comment (banks) [reply config.go:234]: Nope, good catch. I thought I'd fixed this but only in one of the places!

Comment (banks) [inline config.go:234]:
```suggestion
	// send an entire snapshot. The value passed here updates the setting at runtime
	// which will take effect as soon as the next snapshot completes and truncation
	// occurs.
```

File: snapshot.go
Comment (banks) [reply snapshot.go:223]:
```suggestion
	// Use a consistent value for trailingLogs for the duration of this method
```

**Comment by banks** [comment]
Thanks for the reviews folks! Waiting for CI to be green after typo fixes and I'll merge.

**Pull request merged**
